Custom realm emoji

blablacio opened the pull request:
Adding some functionality to allow for managing custom realm emoji from admin.

smarx commented:
*Automated message from Dropbox CLA bot*

@blablacio, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented on zerver/models.py:
```diff
@@ -214,8 +215,8 @@ def remote_user_to_email(remote_user):
-    name = models.TextField()
-    img_url = models.TextField()
+    name = models.TextField(validators=[MinLengthValidator(1), RegexValidator(regex=r'^[0-9a-zA-Z_\-\.]*$')])
+    img_url = models.URLField()
```
What happens to existing instances of a model when you change the type of a field from TextField to URLField?  Seems like this may involve a messy migration.

blablacio replied:
Well, on the database the field is changed from `text` to `character varying(200)` as URLField enforces a `max_length=200` by default. Other than that there should be no problems as values are still retained, even if they don't represent valid URLs.

timabbott replied:
Interesting, so you're saying that either Django or Postgres retains the value in the existing column even when its type is changed from `text` to `varchar(200)`?  Seems worth testing.

blablacio replied:
Yes, that's the case here in particular and in most (if not all) cases when changing from `text` to `character varying` and vice versa. I think it would only be a problem if we already have URLs longer than 200 characters in existing databases. If you think this is worth testing or even extending the migration in order to handle such cases, let me know.

Maybe I should rather set a bigger `max_length` to the URLField in order to avoid such problems? A perfect solution would be to also check the existing URLs in the database against the URLValidator that Django provides.

timabbott replied:
I think there aren't a ton of these in existence, and a 200 character URL is pretty long, so I think it's OK -- just wanted to make sure we're not setting ourselves up for an unpleasant migration.

timabbott commented on zerver/views/__init__.py:
```diff
@@ -1291,3 +1292,23 @@ def email_unsubscribe(request, type, token):
+        do_add_realm_emoji(user_profile.realm, emoji_name, emoji_url)
+    except ValidationError as e:
+        if '__all__' in e.message_dict:
+            return json_error(e.message_dict['__all__'])
```
When would this case happen?

blablacio replied:
Sorry, missed that one. This happens when you try to add an emoji with a duplicate name, the error looks like: `Realm emoji with this Realm and Name already exists.`

timabbott commented on zerver/views/__init__.py:
```diff
@@ -1291,3 +1292,23 @@ def email_unsubscribe(request, type, token):
+    except ValidationError as e:
+        if '__all__' in e.message_dict:
+            return json_error(e.message_dict['__all__'])
+        return json_error('Enter a valid name and URL')
```
Can we make the error message here more clear about what's wrong the name and/or URL?

timabbott commented on zerver/views/__init__.py:
```diff
@@ -1291,3 +1292,23 @@ def email_unsubscribe(request, type, token):
+
+def delete_emoji(request, user_profile, emoji_name):
+    do_remove_realm_emoji(user_profile.realm, emoji_name)
+    return json_success({})
```
I think probably it makes sense to put these new routes in a new zerver/views/realm_emoji.py file, given that it's we're splitting up the view files.

Also, fix the missing newline at the end of the file.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2883,7 +2883,9 @@ def notify_realm_emoji(realm):
 def do_add_realm_emoji(realm, name, img_url):
-    RealmEmoji(realm=realm, name=name, img_url=img_url).save()
+    emoji = RealmEmoji(realm=realm, name=name, img_url=img_url)
+    emoji.full_clean()
```
Hmm, we haven't used `full_clean` before; I'll need to look into that feature...

blablacio replied:
`full_clean` is needed to run validators on the model field. Forms call it for example, but `save()` method does not. Here's an old thread that is still valid: https://groups.google.com/forum/#!topic/django-developers/uIhzSwWHj4c

timabbott replied:
So, the main problem I have with this pattern is that the `do_foo` functions in this file all have the semantics of executing a particular change under the assumption that all validation has already been done.  Here, we're adding a function using that same naming pattern, but that does the validation itself (rather than expecting the validation to be done in the caller).  So I think we should at least rename this function to `check_add_realm_emoji` for clarity.  

Finally, do you think there should be any validation that the URL actually loads?

blablacio replied:
I have not enabled URL checks as this is wasting resources on our side and I think that the user should make sure the URL he adds is the right one and it loads. If you think otherwise, please let me know and I'll enable this check.

timabbott replied:
Yeah that sounds reasonable -- now that we have a preview, it'll be obvious to the user if the URL is busted.

timabbott commented on static/js/admin.js:
```diff
@@ -408,6 +432,51 @@ exports.setup_page = function () {
+
+        channel.put({
+            url: "/json/realm/emoji",
+            data: $(this).serialize(),
```
I'm a little confused by how this code is using `$(this)`, can you explain it a bit?

blablacio replied:
`this` corresponds to the form in this context. It usually yields the element you're binding.
An excerpt from https://api.jquery.com/on/:

> When jQuery calls a handler, the this keyword is a reference to the element where the event is being delivered; for directly bound events this is the element where the event was attached and for delegated events this is an element matching selector. (Note that this may not be equal to event.target if the event has bubbled from a descendant element.) To create a jQuery object from the element so that it can be used with jQuery methods, use $( this ).

timabbott commented on static/styles/zulip.css:
```diff
@@ -4183,3 +4184,12 @@ li.show-more-private-messages a {
+
+.admin_emoji_table caption {
+    font-size: 18px;
+}
```
Fix the no-newline-at-end-of-file issue marked here.

timabbott commented on static/templates/admin_tab.handlebars:
```diff
@@ -40,87 +40,111 @@
+                <th>Actions</th>
+              </tbody>
+            </table>
+            <div id="admin_page_bots_loading_indicator"></div>
```
Can you rebase so that you have the reindentation of the other sections of this page that aren't realm_emoji related as its own commit?  I'd like that to be separate from the new feature code, as per http://zulip.readthedocs.org/en/latest/code-style.html#version-control.

timabbott commented:
Thanks for working on this @blablacio!

I posted a bunch of comments (mostly on small style issues); you'll also want to fix the lint/test errors.

blablacio commented:
Fixed issues and squashed commits. There was a problem with the lint script, I'll push a fix in the `openbsd-compat` branch. I'll also commit the indentation changes separately in master.

timabbott commented on zerver/models.py:
```diff
@@ -214,8 +215,8 @@ def remote_user_to_email(remote_user):
     realm = models.ForeignKey(Realm)
-    name = models.TextField()
-    img_url = models.TextField()
+    name = models.TextField(validators=[MinLengthValidator(1), RegexValidator(regex=r'^[0-9a-zA-Z_\-\.]*$')])
```
Can you add a comment describing the logic behind why this is the right Regex to user here?

blablacio replied:
I will add a comment, even though I'm not sure this is the best regex. That's what we initially discussed, but I think it doesn't make much sense to allow punctuation at the end of the name and I think that would cause some issues with the requests in the frontend too (I think we don't encode the name of the emoji when deleting).

timabbott replied:
OK, probably makes sense to harden the regex then.  Ideally we wouldn't accept names that are going to cause problems with the frontend (though arguably that should be fixed by encoding the name properly?).

timabbott commented on static/js/admin.js:
```diff
@@ -110,6 +127,15 @@ exports.setup_page = function () {
+        timeout: 10 * 1000,
+        success: populate_emoji,
+        error: failed_listing_emoji
+    });
```
I think we don't need this query -- the realm emoji are already sent down to the browser client for all users (admin or no) in `page_params.realm_emoji` and processed in static/js/emoji.js (and updated as they are changed by the `realm_emoji` event in server_events.js).

blablacio replied:
Yes, that's right, but I though this might need to be accessible through the API anyway, so I added it as an API endpoint. Should I just remove the query in the frontend or entirely remove this endpoint?

timabbott replied:
Makes sense.  It's probably reasonable to just remove it from the frontend.

timabbott commented:
I took a deeper look at the implementation and posted a few comments; I also played around with the feature, and have a few thoughts on how to improve up the design (screenshot here for others following along).

![image](https://cloud.githubusercontent.com/assets/2746074/13942468/035dbf8a-efb0-11e5-8726-cc48ea2fdea2.png)

* We should display the existing custom emoji at a fixed size with the same dimensions ratio as what you'll get as the actual emoji (it's possible we should to show it at e.g. 2-4x magnification from how it'll appear in messages, so it's easier to see what you'll actually get?)
* The spacing of the columns looks kinda funny -- if we add a fixed display size for the emoji I'd probably make the columns sit more tightly next to each other rather than having a huge space between the second and third columns.
* I think it'd be good to have more of a separation from the previous section.  I think the mechanism we used pre 62d021f3990ae5bee832ddd527cf9cce01441977 of making a new gray block to set aside the section with e.g. "Custom emoji" as the heading and a nice icon would work pretty well.
* Adding more visual offset for the "create a new one" section, similar to how the bots UI on /#settings would I think also make it more clear what's going on.

What do you think of those ideas @blablacio ?

blablacio commented:
Added Node tests in `templates.js`, so build should be working now.

blablacio commented:
You comments about the visual improvements are spot on, will make the needed changes.

blablacio commented:
Just addressed style issues, here's a preview:
![emoji](https://cloud.githubusercontent.com/assets/3009538/13999035/0154e120-f143-11e5-9021-56a4beb2384d.png)

timabbott commented:
​As a quick thought, I think we should have the 3 columns of the table next
to each other rather than having that huge space between "name" and "image"

blablacio commented:
@timabbott So, basically leaving the empty space after the third column?

timabbott commented:
Yeah.  I think it might make sense to also think about making the overall widget have more of a margin on the left side (though if we do that we should do something that looks consistent with the other pages, so probably that would mean adding margin for this category of widget; looking at the other similar pages, I think it'd probably make them all look better)...

timabbott commented:
(And sorry for the slow reply!  Last week was very busy)

timabbott commented on static/styles/zulip.css:
```diff
@@ -4184,3 +4200,17 @@ li.show-more-private-messages a {
+.admin_emoji_table th.image {width: 80px;}
+
+.emoji_image {width: 50px; display: block;}
+.emoji_image img {max-width: 100%;}
```
I think stylistically our CSS always has a newline after the `{`, so probably should do that here too for consistency.

timabbott commented:
@blablacio it'd be great if you could focus on getting this one finished, since I think it's almost done!

blablacio commented:
Sorry for the late response, last couple of weeks were really busy. All issues have been addressed and code rebased, you can take a look at the emoji table and let me know if I should wrap the tables on other admin tabs similarly.

timabbott commented:
Cool, thanks @blablacio!  The new changes are definitely an improvement, and I think the new admin page form looks pretty good.

I noticed this issue while playing with this:
* If one adds a realm emoji with a capital letter in the name (e.g. "Cowboy"), I can't get the new realm emoji to show up in the autocomplete in the compose box, even after reloading -- I'm not sure what the issue is there but it seems worth debugging.  Seems likely to have been a preexisting bug.

timabbott commented on static/js/admin.js:
```diff
@@ -408,6 +428,67 @@ exports.setup_page = function () {
+                        $("<p>").addClass("text-error").text($.parseJSON(xhr.responseText).msg)
+                    );
+                } else {
+                     btn.text("Failed!");
```
Would it make sense for this to use Zulip's standard `ui.report_error` mechanism?

blablacio replied:
It would probably make sense, but I think this is the "canonical" way to do it (at least looking at admin.js). It sounds reasonable to me as we denote the table row within which the error occurs. I could change it if you think it makes sense, but I guess it would be a good idea to change the other few occurrences of this pattern in admin.js too.

timabbott replied:
Hmm, so it is.  We should probably switch these but maybe makes sense to do in a separate commit / PR -- doesn't seem super important.  Can you open an issue and label it with "refactoring"?

blablacio replied:
Sure, will do.

timabbott commented on static/styles/zulip.css:
```diff
@@ -4217,3 +4233,18 @@ li.show-more-private-messages a {
+}
+
+.emoji_image {width: 50px; display: block;}
+.emoji_image img {max-width: 100%;}
```
These should be in the same multi-line CSS style as the rest of the file

timabbott commented on zproject/urls.py:
```diff
@@ -188,7 +188,12 @@
+    url(r'^realm/emoji$', 'rest_dispatch',
+        {'GET': 'list_emoji',
+         'PUT': 'upload_emoji'}),
+    url(r'^realm/emoji/(?P<emoji_name>[a-zA-Z0-9.-]+)$', 'rest_dispatch',
```
So this forbids `_` while the `zerver/models.py` implementation allows it -- is that a bug here?

timabbott commented on zerver/models.py:
```diff
@@ -214,8 +215,9 @@ def remote_user_to_email(remote_user):
-    name = models.TextField()
-    img_url = models.TextField()
+    name = models.TextField(validators=[MinLengthValidator(1),
+                                        RegexValidator(regex=r'^[0-9a-zA-Z_\-\.]*(?<![.\-_])$')])
```
Maybe add a comment explaining the second part of this regexp is enforcing that the name not end with a punctuation character?

timabbott commented on zerver/migrations/0014_realmemoji_name_regex.py:
```diff
@@ -0,0 +1,20 @@
+            name='name',
+            field=models.TextField(validators=[django.core.validators.MinLengthValidator(1), django.core.validators.RegexValidator(regex=b'^[0-9a-zA-Z_\\-\\.]*(?<![.\\-_])$')]),
+        ),
+    ]
```
I think for these migrations it'd probably be cleaner to just have a single file 0013_realmemoji.py depending on 0012 and containing the code for the 0010 + 0014 added in this PR; there aren't many other database migration commits in flight.  Probably easiest way to do this would be to just delete these migrations and rerun makemigrations.

timabbott commented:
Oh one more note on the Emoji creation form: we should probably give a wider text box for the URL field since many URLs will probably not fit there.

blablacio commented:
Okay, I'll investigate what is the case with the capital letters in the name and address the other issues you've pointed.

blablacio commented:
Okay, so here's the deal with the capital letters: in composebox_typeahead.js we have `query_matches_emoji` function that converts input to lower case like this `return (emoji.emoji_name.indexOf(query.toLowerCase()) !== -1);`

If I remove the `toLowerCase()` call, it works as expected, but I'm not sure if that's going to break anything. Any ideas about the purpose of converting to lower case?

timabbott commented:
My guess is that code's purpose is to make emoji names case-insensitive, which is probably a pretty reasonable way for them to work.  Maybe try making it `emoji.emoji_name.toLowerCase().indexOf(query.toLowerCase())` ?

blablacio commented:
Yep, that would make sense. Will change it as suggested and run tests.

blablacio commented:
I fixed all the stuff, but would like you to okay the width change of the URL field as I had to tweak the whole form a bit (otherwise the label would stick on one row and the input on the next one). Attaching a preview here:
![emoji_form](https://cloud.githubusercontent.com/assets/3009538/14831141/f5c19e14-0bfc-11e6-9937-abdca5614e7d.png)

timabbott commented:
Yeah that seems reasonable

blablacio commented:
All done, I unified the regexes in models and urls too.

timabbott commented:
Great!  Looks like there are some issues with narrow screens; I think it's reasonable for the text box to shrink in that case but probably not great for it to stick out over the border area.

![screenshot from 2016-04-26 12-40-36](https://cloud.githubusercontent.com/assets/2746074/14831813/1ae6f604-0bac-11e6-81fa-272388454eef.png)

Oh, one more thing -- ideally, the Zulip tests would work without Internet access, so maybe the example image URL used in the tests should be something included in the Zulip distribution (e.g. a "jenkins" emoji that points to /static/images/integrations/logos/jenkins.png).

blablacio commented:
Okay, I'll set the input width in percents, so it shrinks with the screen size. I'll also look into making tests run without an internet connection.

blablacio commented:
Done.

timabbott commented:
Probably should rename `MouseFace` to `jenkins` in the test?

timabbott commented:
Otherwise lgtm, I'll just fix that and merge.  Thanks @blablacio!

timabbott commented:
I split out f5fe2d4bf776c027f5ea5b6ef99481ffcf0c70e7 into its own commit, since it's in some ways an unrelated bug fix, and then merged this.  Thanks @blablacio!
